plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,160 @@
+import           Distribution.Compat.Prelude            (Generic)
+import           Distribution.PackageDescription        (packageDescription,
+                                                         specVersion)
+import           System.FilePath                        (dropFileName,
+                                                         splitPath,
+                                                         takeExtension, (</>))
+import           Text.Regex.TDFA
+
+
+-- | Given a path to a haskell file, finds all cabal files paths
+--   sorted from the closest to the farthest.
+--   Gives all found paths all the way to the root directory.
+findResponsibleCabalFile :: FilePath -> IO [FilePath]
+findResponsibleCabalFile uriPath = do
+  contents <- mapM listDirectory allDirPaths
+  let filesWithPaths = concat $ zipWith (\path content -> map (path </>) content) allDirPaths contents
+  let cabalFiles = filter (\c -> takeExtension c == ".cabal") filesWithPaths
+  pure $ reverse cabalFiles -- sorted from closest to the uriPath
+  where dirPath = dropFileName uriPath
+        allDirPaths = scanl1 (</>) (splitPath dirPath)
+
+
+-- | Gives a code action that calls the command,
+--   if a suggestion for a missing dependency is found.
+--   Disabled action if no cabal files given.
+missingDependenciesAction :: PluginId -> Int -> Uri -> Diagnostic -> [FilePath] -> [CodeAction]
+missingDependenciesAction plId maxCompletions _ diag cabalFiles =
+  case cabalFiles of
+    [] -> [CodeAction "No .cabal file found" (Just CodeActionKind_QuickFix) (Just []) Nothing
+                                             (Just (CodeActionDisabled "No .cabal file found")) Nothing Nothing Nothing]
+    (cabalFile:_) -> mkCodeAction cabalFile <$> missingDependenciesSuggestion maxCompletions (_message diag)
+  where
+    mkCodeAction cabalFile (suggestedDep, suggestedVersion) =
+      let
+        versionTitle = if T.null suggestedVersion then T.empty else "version " <> suggestedVersion
+        title = "Add dependency " <> suggestedDep <> " " <> versionTitle
+
+        version = if T.null suggestedVersion then Nothing else Just suggestedVersion
+        params = CabalAddCommandParams {cabalPath = cabalFile, dependency = suggestedDep, version=version}
+        command = mkLspCommand plId (CommandId cabalAddCommand) "Execute Code Action" (Just [toJSON params])
+      in CodeAction title (Just CodeActionKind_QuickFix) (Just []) Nothing Nothing Nothing (Just command) Nothing
+
+-- | Gives a mentioned number of hidden packages given
+--   a specific error message
+missingDependenciesSuggestion :: Int -> T.Text -> [(T.Text, T.Text)]
+missingDependenciesSuggestion maxCompletions msg = take maxCompletions $ getMatch (msg =~ regex)
+  where
+    regex :: T.Text -- TODO: Support multiple packages suggestion
+    regex = "Could not load module \8216.*\8217.\nIt is a member of the hidden package \8216([a-z]+)[-]?([0-9\\.]*)\8217"
```

Comment:
Depending on environment variables, GHC can use `\8216...\8217`, or simply `'...'`. It would be prudent to match on both.

Reply:
Ok, I will fix that. Is there a way to alter the environment to test the second case? On my system, I get the first one by default.

Reply:
Try running VSCode with the environment variable `GHC_NO_UNICODE=1`.

See https://github.com/haskell/haskell-language-server/issues/3841#issuecomment-1762048408